keps/sig-auth/3766-referencegrant/README.md

```diff
@@ -0,0 +1,740 @@
+
+#### Cross-Namespace References may Weaken Namespace Boundaries
+Although we believe that the handshake required for cross-namespace references
+with ReferenceGrant ensures these references will be safe, it does potentially
+weaken existing namespace boundaries. We believe ReferenceGrant will have a net
+benefit on the ecosystem as it will allow workloads, secrets, and configuration
+to be deployed in separate namespaces that more clearly match up with desired
+authorization.
+
+## Design Details
+
+### General Notes
+
+#### `ReferenceGrant` is half of a handshake
+
+When thinking about ReferenceGrant, it is important to remember that it does not
+do anything by itself. It *Grants* the *possibility* of making a *Reference*
+across namespaces. It's intended that _another object_ (that is, the From object)
+complete the handshake by creating a reference to the referent object (the To
+object).
+
+#### `Resource` vs `Kind`
+
+When creating a metaresource (that is, a resource that targets other resources)
+like ReferenceGrant, it's important to consider if the metaresource uses the more
+common `Kind` or the more correct `Resource`.
+
+In the original Gateway API implementation, we chose to use `Kind` rather than
+`Resource`, mainly to improve the user experience. That is, it's easier users
+to take the value from the `kind` field at the top of the YAML they are already
+using, and put it straight into these fields, rather than needing to do a
+kind-resource lookup for every user's interaction with the API. @robscott even
+ended up making https://github.com/kubernetes/community/pull/5973 to clarify
+the API conventions.
+
+However, in discussion on this KEP, it's clear that the more generic nature of
+_this_ API requires the additional specificity that `Resource` provides.
+
+The Gateway API ReferenceGrant looked like this:
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: ReferenceGrant
+metadata:
+  name: allow-gateways
+  namespace: bar
+spec:
+  from:
+    # Note that in Gateway API, Group is currently defaulted
+    # to this, which means you to explicitly set the group to
+    # the empty string for Core resources. We should definitely
+    # change this.
+    - group: "gateway.networking.kubernetes.io"
+      kind: Gateway
+      namespace: foo
+  to:
+   - group: ""
+     kind: Secret
+```
+
+The new version will look like this instead:
+```yaml
+apiVersion: authorization.k8s.io/v1alpha1
+kind: ReferenceGrant
+metadata:
+  name: allow-gateways
+  namespace: bar
+spec:
+  from:
+    # Assuming that we leave the default for Group to the empty
+    # string, so that Core objects don't need additional config.
+    - group: "gateway.networking.kubernetes.io" 
+      resource: gateways
+      namespace: foo
+  to:
+    - resource: secrets
+
+```
+
+The new version communicates the scope more clearly because `resource` is
+unambiguous and corresponds to exactly one URL path on the API Server.
+
```

Comment:
```suggestion

This change also leaves room for an enhancement.
Whether we have an in-tree or CRD implementation, we can rely on the exact matching that
the plural resource name gives us, and
[warn](https://kubernetes.io/blog/2020/09/03/warnings/) if either side of the grant is for
an API that's not served by this cluster.
```

Depending on what the eventual scope of issue 3766 becomes, we might  one day also add matching on
`uid` (in case the specific UID matters).

Reply:
Updated, thanks!